server/src/main/java/org/elasticsearch/common/time/JavaDateFormatter.java

```diff
@@ -68,24 +67,16 @@ class JavaDateFormatter implements DateFormatter {
     };
 
     private final String format;
-    private final DateTimeFormatter printer;
-    private final DateTimeFormatter[] parsers;
-    private final RoundUpFormatter roundupParser;
+    private final DateTimePrinter printer;
+    private final DateTimeParser[] parsers;
+    final DateTimeParser[] roundupParsers;
 
-    private static final class RoundUpFormatter extends JavaDateFormatter {
-
-        RoundUpFormatter(String format, DateTimeFormatter[] roundUpParsers) {
-            super(format, roundUpParsers[0], (RoundUpFormatter) null, roundUpParsers);
-        }
-
-        @Override
-        JavaDateFormatter getRoundupParser() {
-            throw new UnsupportedOperationException("RoundUpFormatter does not have another roundUpFormatter");
-        }
+    JavaDateFormatter(String format, DateTimeFormatter formatter) {
+        this(format, new JavaTimeDateTimePrinter(formatter), new JavaTimeDateTimeParser(formatter));
     }
 
     // named formatters use default roundUpParser
-    JavaDateFormatter(String format, DateTimeFormatter printer, DateTimeFormatter... parsers) {
+    JavaDateFormatter(String format, DateTimePrinter printer, DateTimeParser... parsers) {
```

Comment:
Could we avoid the massive changes to the constants in this PR by maintaining an alternate constructor that still takes DateTimeFormatter, which then wraps the formatters and delegates to this ctor?

Reply:
I've kept the single-formatter ctor when the same object is used for both, but when there's separate formatters for printing & parsing I thought it best it was explicit which one is which

Reply:
I've created a static method which keeps it out of `JavaDateFormatter` but reduces the churn

Reply:
I'm still seeing a lot of churn in this PR. 700+ lines in DateFormatters, which is more than half the PR, and it's just boilerplate (wrapping).

Explicit is good, but if it comes at the cost of readability (ie boilerplate) I'm not sure it is better. I don't follow how changing the signature here is necessary for having additional printer/parser implementations, can you elaborate?

Reply:
It's a minor point within the PR, but here goes...

As part of this PR, I'm generally improving the code in these classes. First, lets look at the previous constructor signature:

    JavaDateFormatter(String, DateTimeFormatter, DateTimeFormatter...)

There's two `DateTimeFormatter`-type arguments here. The first is for the printer, the second for the parsers. In contrast to methods like `EnumSet(E, E...)`, the two arguments are _not_ doing the same thing. This leads to ambiguity, as you'll end up with a method call like `new JavaDateFormatter(str, formatter1, formatter2, formatter3)`, with no indication that `formatter1` and `formatter2`/`formatter3` are actually doing different things. However, given `DateTimeFormatter` is the only type to use here, there's not much else we could have done. Additionally, the previous code allowed the printer formatter to also be used as a parser, if no parsers were specified, making the single-arg array-arg at least justifiable. Still not great though.

With this PR, this introduces separate types for the printer and parser types. This changes the core constructor signature to

    JavaDateFormatter(String, DateTimePrinter, DateTimeParser...)

Much better. We now have different types for the printer and parser, removing the ambiguity between the uses of the singular arg and the array arg. So the question is - do we keep the existing signature as a delegating constructor? I decided _no_, because we have this opportunity to reduce ambiguity with these new types. This, as you mentioned, caused a lot of churn in the `DateFormatters` class when I added the adapter classes to compensate for the changed signature.

Churn, in an of itself, is not necessarily a problem, if the refactoring improves the quality of the code. But in this case, it does add lots of lines of code to `DateFormatters` with creating the adapter instances. So, to compensate, I added private static methods to `DateFormatters` that adds the adapter classes. This is not as bad as adding constructor overloads to the `JavaDateFormatter` class, as it is entirely contained within `DateFormatters` - it is only used when constructing the static members, in a strictly contained context. So I judged the ambiguity is less of an issue because there's a specific pattern being used here which compensates. This would not necessarily be the case if the signature was a generally accessible constructor on `JavaDateFormatter`.

This is also a refactoring to enable the use of different implementations of the printer and parser interfaces. Within `JavaDateFormatter`, the printer and parser are entirely separate types. Having a constructor in `JavaDateFormatter` that takes a `DateTimeFormatter` for both muddies the distinction between those concepts within that class. Moving the adapter construction to `DateFormatters` ensures the implementations of printing and parsing using the adapters for `DateTimeFormatter` are only contained within `DateFormatters` - `JavaDateFormatter` does not care what the implementations are, only that they are provided. So then, later on, adding new implementations of the printing & parsing interfaces and using them is only a logical change for `DateFormatters`, the calls to `JavaDateFormatter` do not change at all. If the adapters were created by `JavaDateFormatter`, then the constructor calls performed by `DateFormatters` _would_ change when new implementations are used. This is not something that should modify how `JavaDateFormatter` is used.

That is what was going through my head when deciding how to refactor these classes, and where the adapters and arguments belonged.